create documentation for operations using the new Beagle dsl (#767)
ZupIT/beagle-docs: content/pt/resources/dsl-operations/how-to-register-a-new-operation.md

---
title: Como registrar uma nova operação
weight: 375

> description: "Nesta seção, você encontra descrição completa das Custom Operations."

Reviewer: ```suggestion
description: "Nesta seção, você encontra como registrar uma nova operação."
```

---

---

> Depois que você viu que é possível realizar[ **Operações**]({{< ref path="/resources/dsl-operations" lang="pt" >}}) do tipo soma, subtração, etc, utilizando o contexto e a nova DSL, você também pode criar a sua própria operação!

Reviewer: ```suggestion
No Beagle é possível realizar [ **Operações**]({{< ref path="/resources/dsl-operations" lang="pt" >}}) do tipo soma, subtração, etc, utilizando o contexto e a DSL, mas você também pode criar a sua própria operação ou 'Custom Operation'.
```

Para registrar sua operação customizada, é necessário usar o método **createOperation** que recebe como parâmetros:
- Uma `String` com o nome da operação;
- Um `Array<out Any?>`, contendo os parâmetros necessários de acordo com a operação customizada.

## Exemplo

> Veja abaixo o exemplo utilizando a operação `isValidCpf` onde o texto do componente `Text` varia de acordo com o resultado da verificação:

Reviewer: ```suggestion
O exemplo abaixo mostra a operação `isValidCpf` e o texto do componente `Text` varia de acordo com o resultado da verificação, veja como:
```

```
fun isValidCpf(param: Array<out String?>) : Bind.Expression<Boolean> {
    param.filterNotNull().map { value ->
        if (value.isEmpty()) return expressionOf("@{${false}}")
        val numbers = arrayListOf<Int>()
        value.filter { it.isDigit() }.forEach {
            numbers.add(it.toString().toInt())
        }
        if (numbers.size != 11) return expressionOf("@{${false}}")
        for (n in 0..9) {
            val digits = arrayListOf<Int>()
            repeat((0..10).count()) { digits.add(n) }
            if (numbers == digits) return expressionOf("@{${false}}")
        }
        val dv1 = ((0..8).sumOf { (it + 1) * numbers[it] }).rem(11).let {
            if (it >= 10) 0 else it
        }

        val dv2 = ((0..8).sumOf { it * numbers[it] }.let { (it + (dv1 * 9)).rem(11) }).let {
            if (it >= 10) 0 else it
        }

        return expressionOf("@{${numbers[9] == dv1 && numbers[10] == dv2}}")
    }
    return createOperation("isValidCpf", arrayOf(param))
}
```

> No exemplo acima, foi criado uma operação customizada que valida um Cpf e retorna se o resultado foi válido ou não.
> Para isso, foram feitas as validações necessárias e retornou o método **createOperation** com o nome da operação e o parâmetro esperado. Com isso, foi obtido o seguinte resultado:

Reviewer: ```suggestion
No exemplo acima a operação customizada criada valida um CPF e retorna se o resultado foi válido ou não.
Para isso, as validações necessárias foram feitas e depois o método **createOperation** retornou com o nome da operação e o parâmetro esperado. O resultado foi:
```
